### ISSUE TITLE
Invalid suggestion of `or_fun_call` causes a compiler error `error[E0277]: expected a `FnOnce<()>` closure, found `Vec<_>``

### ISSUE THREAD

**Issue by mgacek8** [rust-lang/rust-clippy#6748]
Lint name: or_fun_call

I tried this code:

```rust
use std::collections::HashMap;
fn main() {
    let mut hash_map: HashMap<_, Vec<usize>> = HashMap::new();
    let _ = hash_map.entry("test".to_owned()).or_insert(vec![]);

    println!("hash_map: {:?}", hash_map);
}

```

I expected to see this happen: suggestion to use `or_insert_with(|| vec![])` instead of `or_insert(vec![])`

Instead, this happened:  clippy suggested me to use  `or_insert_with(vec![])`
```rust
warning: use of `or_insert` followed by a function call
 --> src/main.rs:4:47
  |
4 |     let _ = hash_map.entry("test".to_owned()).or_insert(vec![]);
  |                                               ^^^^^^^^^^^^^^^^^ help: try this: `or_insert_with(vec![])`
  |
  = note: `#[warn(clippy::or_fun_call)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call

warning: 1 warning emitted
```
and that suggestion causes the compiler error.
```rust
error[E0277]: expected a `FnOnce<()>` closure, found `Vec<_>`
 --> src/main.rs:4:62
  |
4 |     let _ = hash_map.entry("test".to_owned()).or_insert_with(vec![]);
  |                                                              ^^^^^^ expected an `FnOnce<()>` closure, found `Vec<_>`
  |
  = help: the trait `FnOnce<()>` is not implemented for `Vec<_>`
  = note: wrap the `Vec<_>` in a closure with no arguments: `|| { /* code */ }`
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```
Playground [link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8aff832294ecb20bff47a1267b490ded)

I found that this only happens when the vector is empty.
When it is not, the suggestion is correct:
```rust
4 |     let _ = hash_map.entry("test".to_owned()).or_insert(vec![123]);
  |                                               ^^^^^^^^^^^^^^^^^^^^ help: try this: `or_insert_with(|| vec![123])`
 ```

### Meta

- clippy 0.1.52 (5fa22fe 2021-02-14)
- rustc stable 1.50.0

**Comment by camsteffen** [comment]
Could be fixed with #3812.

**Comment by mgacek8** [comment]
Shouldn't this lint for `.or_insert(vec![])` suggest using `.or_insert_with(Vec::new)` (instead of `.or_insert_with(vec![])` that causes a compiler error)? 
And then, as in #3812, there could be another link that suggest `std::collections::btree_map::Entry::or_default` instead of  `.or_insert_with(Vec::new)`.
Please let me know if it makes sense to fix that (`.or_insert(vec![])` -> `.or_insert_with(Vec::new)`) . If it does I'd like to do it.

**Comment by camsteffen** [comment]
We should have both fixed actually, since `vec![]` could be any macro that calls any function. When #3812 is fixed, it should just suggest `or_default()` instead of `vec![]`.

So @mgacek8 if you would like to just fix the this issue that would be great!
